flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTypeUtil.java

```diff
@@ -256,9 +258,9 @@ private static DataType toFlinkPrimitiveType(PrimitiveTypeInfo hiveType) {
 			case DOUBLE:
 				return DataTypes.DOUBLE();
 			case DATE:
-				return DataTypes.DATE();
+				return DataTypes.DATE().bridgedTo(Date.class);
 			case TIMESTAMP:
-				return DataTypes.TIMESTAMP();
+				return DataTypes.TIMESTAMP(3).bridgedTo(Timestamp.class);
```

Comment:
`java.sql.Timestamp` record time in milliseconds, which is 10^{-3} seconds.

Reply:
IIRC, `java.sql.Timestamp` has a separate field to store nanoseconds and supports precision 9. E.g. you can create a Timestamp like `Timestamp.valueOf('2019-08-06 14:00:00.123456789')`

Reply:
Flink previously stores timestamp using a long variable, so only a timestamp of precision 3 can be converted to the `SqlTimeInfo.TIMESTAMP_INFO`.

Reply:
If that's the case, please add some comments here to explain why the precision has to be 3. Besides, does this mean we'll lose precision when reading a timestamp from Hive into Flink?

Reply:
Yes, precision will be lost. See `TimestampConverter` in `DataTypeConverters`.

Reply:
@twalthr Do you have any comment on this loss of precision in conversion? Thanks.

Reply:
Precision from Hive should be forwarded to Flink. It should be the responsibility of the planner to throw an exception if it does not support the given precision of a source.

Reply:
+1 to keep precision and throw exception.
Indeed there are somewhere to loss precision because of conversion between TypeInformation and DataType. But we do our best to use DataType and keep precision. And we just convert the timestamp with precision smaller than 3. (`LegacyTypeInfoDataTypeConverter.canConvertToTimestampTypeInfoLenient`)

Reply:
Hive's timestamp has precision 9. If Flink internally stores `java.sql.Timestamp` as a long (like @TsReaper mentioned in `TimestampConverter`), seems we cannot avoid the loss. Does Flink support precision 9 with `java.time.LocalDateTime`?

Reply:
Only support precision smaller(or equal) than 3.

Reply:
Thanks for the discussion on the precision issue. As the default precision in hive is 9, the connector will be literally not functional if we directly throws an exception. So I think it will be better to log a warning message and add in the connector's document that timestamp precision may be lost.

Reply:
I agree it's better to warn and document this limitation than totally ban timestamp from Hive connector. And in the long run, shall we support timestamp with higher precision in Flink? Since both `java.sql.Timestamp` and `java.time.LocalDateTime` support up to precision 9, I don't see a good reason why Flink has to limit precision to 3.